R/join-cols-compat.R

```diff
@@ -0,0 +1,175 @@
+join_cols <- function(x_names,
+                      y_names,
+                      by,
+                      ...,
+                      suffix = c(".x", ".y"),
+                      keep = NULL,
+                      error_call = caller_env()) {
+  check_dots_empty0(...)
+
+  if (is_false(keep) && any(by$condition != "==")) {
+    abort(
+      "Can't set `keep = FALSE` when using an inequality, rolling, or overlap join.",
+      call = error_call
+    )
+  }
+
+  check_duplicate_vars(x_names, "x", error_call = error_call)
+  check_duplicate_vars(y_names, "y", error_call = error_call)
+
+  check_join_vars(by$x, x_names, by$condition, "x", error_call = error_call)
+  check_join_vars(by$y, y_names, by$condition, "y", error_call = error_call)
+
+  suffix <- standardise_join_suffix(suffix, error_call = error_call)
+
+  x_by <- set_names(match(by$x, x_names), by$x)
+  y_by <- set_names(match(by$y, y_names), by$y)
+
+  x_loc <- seq_along(x_names)
+  names(x_loc) <- x_names
+  if (is_null(keep)) {
+    # In x_out, equi key variables need to keep the same name, and non-equi
+    # key variables and aux variables need suffixes for duplicates that appear
+    # in y_out. This is equivalent to `keep = TRUE` for the non-equi keys and
+    # `keep = FALSE` for the equi keys.
+    equi <- by$condition == "=="
+    y_aux <- setdiff(y_names, c(by$x[equi], by$y[equi]))
+    x_ignore <- by$x[equi]
+    x_check <- !x_names %in% x_ignore
+    names(x_loc)[x_check] <- add_suffixes(x_names[x_check], c(x_ignore, y_aux), suffix$x)
+  } else if (is_false(keep)) {
+    # In x_out, key variables need to keep the same name, and aux
+    # variables need suffixes for duplicates that appear in y_out
+    y_aux <- setdiff(y_names, c(by$x, by$y))
+    x_ignore <- by$x
+    x_check <- !x_names %in% x_ignore
+    names(x_loc)[x_check] <- add_suffixes(x_names[x_check], c(x_ignore, y_aux), suffix$x)
+  } else {
+    # In x_out, key variables and aux variables need suffixes
+    # for duplicates that appear in y_out
+    names(x_loc) <- add_suffixes(x_names, y_names, suffix$x)
+  }
+
+  y_loc <- seq_along(y_names)
+  names(y_loc) <- add_suffixes(y_names, x_names, suffix$y)
+  if (is_null(keep)) {
+    equi <- by$condition == "=="
+    y_ignore <- by$y[equi]
+    y_loc <- y_loc[!y_names %in% y_ignore]
+  } else if (is_false(keep)) {
+    y_ignore <- by$y
+    y_loc <- y_loc[!y_names %in% y_ignore]
+  }
+
+  # key = named locations to use for matching
+  # out = named locations to use in output
+  list(
+    x = list(key = x_by, out = x_loc),
+    y = list(key = y_by, out = y_loc)
+  )
+}
+
+check_join_vars <- function(vars,
+                            names,
+                            condition,
+                            input,
+                            ...,
+                            error_call = caller_env()) {
+  check_dots_empty0(...)
+
+  if (!is.character(vars)) {
+    message <- glue("Join columns in `{input}` must be character vectors.")
```

Comment:
When does this error arise? Is it an internal error?

Reply:
Probably, it is an internal error. I basically only copied over the file from dplyr https://github.com/tidyverse/dplyr/blob/main/R/join-cols.R.